Rename EmptyTypTree into SyntacticEmptyTypeTree

densh opened the pull request:
1. Change the name as @xeno-by believes that previous name was misleading
2. Remove EmptyTree case as it's not needed any longer, only TypeTrees are used now 

review @xeno-by 
/cc @vladimirnik

densh committed:
Rename EmptyTypTree into SyntacticEmptyTypeTree

1. Change the name as Eugene believes previous name was misleading

2. Remove EmptyTree case as it's not needed any longer

densh commented:
I wonder if we need would also want `tq""` to only match only trees without original. Currently they are just `TypeTree()` patterns without any restrictions.

xeno-by commented:
@densh Absolutely.

xeno-by commented on src/reflect/scala/reflect/internal/BuildUtils.scala:
```diff
@@ -537,13 +537,9 @@ trait BuildUtils { self: SymbolTable =>
-      }
+    // a type tree which wasn't specified by the user but might have been inferred by compiler
+    object SyntacticEmptyTypeTree {
+      def unapply(tt: TypeTree): Boolean = tt.original == null || tt.original.isEmpty
```
What's the difference with `tt.wasEmpty`? /cc @retronym

densh replied:
`tt.wasEmpty` isn't set after parsing and thus can't be used here. @retronym is it meant to be this way?

xeno-by commented:
@densh Do you plan to submit the changes to how `tq""` works as part of this pull request?

densh commented:
@xeno-by yes, will submit them soon today.

densh committed:
Represent tq"" as SyntacticEmptyTypeTree rather than TypeTree()

Such representation codifies the fact that type tree that doesn't have
embedded syntactic equivalent must have been inferred or otherwise
provided by the compiler rather than specified by the end user.

Additionally it also ensures that we can still match trees without
explicit types (e.g. vals without type) after typechecking. Otherwise
the same quote couldn't be used in situations like:

    val q"val x = 42" = typecheck(q"val x = 42")

densh committed:
Better comment for SyntacticEmptyTypeTree

xeno-by commented:
LGTM
